### PR TITLE
DAOS-4419 dfuse: Add container attributes to configure dfuse caching.

### PR DESCRIPTION
**Pull request by ashleypittman**
Set cache parameters as container attributes.
Overhaul the way that caching is handled in dfuse.
Update posix.md with the new attributes.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

File: src/client/dfuse/dfuse_core.c
Comment (ashleypittman) [inline src/client/dfuse/dfuse_core.c:483]:
```suggestion
					       "dfuse-dentry-timeout",
					       "dfuse-ndentry-timeout",
```

File: src/client/dfuse/dfuse_core.c
Comment (ashleypittman) [inline src/client/dfuse/dfuse_core.c:484]: Needs to be boolean on/off

File: src/utils/daos.c
Comment (daltonbohning) [inline src/utils/daos.c:556]: Also need these:
```suggestion
	D_FREE(ap->user);
	D_FREE(ap->group);
}
```

File: src/client/dfuse/dfuse_core.c
Comment (mchaarawi) [inline src/client/dfuse/dfuse_core.c:501]: please add define for size

File: src/client/dfuse/dfuse_core.c
Comment (mchaarawi) [inline src/client/dfuse/dfuse_core.c:520]: it's better to have an strcmp here instead of the name to "dfuse-data-cache"

File: src/client/dfuse/dfuse_core.c
Comment (mchaarawi) [inline src/client/dfuse/dfuse_core.c:545]: same... or a #defined index

File: src/client/dfuse/ops/lookup.c
Comment (mchaarawi) [inline src/client/dfuse/ops/lookup.c:212]: why out_free if alloc failed?

File: src/client/dfuse/ops/setxattr.c
Comment (mchaarawi) [inline src/client/dfuse/ops/setxattr.c:41]: should add a test for this

File: utils/node_local_test.py
Comment (mchaarawi) [inline utils/node_local_test.py:1143]: i would prefer to have all tests with caching enabled, and:
needs_dfuse_with_cache
be
needs_dfuse_without_cache ?

**Review by mchaarawi — CHANGES_REQUESTED** [review]

File: src/client/dfuse/dfuse_core.c
Comment (ashleypittman) [reply src/client/dfuse/dfuse_core.c:520]: Yes, this isn't idea as is.  I'll add a #define index.

File: src/client/dfuse/ops/lookup.c
Comment (ashleypittman) [reply src/client/dfuse/ops/lookup.c:212]: This looks like a bad merge actually, I'll change this bit back.

File: src/client/dfuse/ops/setxattr.c
Comment (ashleypittman) [reply src/client/dfuse/ops/setxattr.c:41]: test_uns_create_with_cache() covers this, without the hook here it creates the UNS, then when you stat it you get the underlying file as it comes from the dentry cache, with this hook it works as expected, so if we hit the theoretical race condition then the test would deadlock.

It's not ideal as I put in the comment and I'm intending to raise this on the fuse mailing list for clarification.

File: src/tests/ftest/cart/util/cart_logparse.py
Comment (ashleypittman) [inline src/tests/ftest/cart/util/cart_logparse.py:483]: This was also included accidentally, I'll remove.

File: utils/node_local_test.py
Comment (ashleypittman) [reply utils/node_local_test.py:1143]: It's really testing two different things, if we enabled caching then a lot of the tests wouldn't be covering what we'd intended them to.

Additionally, there are failures if you just turn on caching, so we probably want some code to run all tests twice, once with and once without, but skip running the once that require caching without caching.

Part of the issue is rename, with caching a simple test ike the following fails:

touch a
mv a b
stat b

What happens is the mv checks to see if b exists for some reason, puts a negative dentry in the cache then the mv doesn't automatically remove the cache entry so the stat fails with "no such file or directory".  I'm still investigating this further but it seems like it should be a kernel bug.

I'll run some more tests on this but I don't want to loose the coverage of running without caching.

File: doc/user/posix.md
Comment (ashleypittman) [inline doc/user/posix.md:175]: typo

Comment (johannlombardi) [inline doc/user/posix.md:175]: typo

File: doc/user/posix.md
Comment (johannlombardi) [inline doc/user/posix.md:173]: Shall we call this section Caching instead since caching is enabled by default now?

File: doc/user/posix.md
Comment (johannlombardi) [inline doc/user/posix.md:217]: I would add a:

!note
     Caching is enabled by default in dfuse. This might cause some parallel applications to fail. Please use disable caching ...

**Review by johannlombardi — CHANGES_REQUESTED** [review]
We should run it with daily_regression too.

File: doc/user/posix.md
Comment (johannlombardi) [reply doc/user/posix.md:217]: Please see below an example of what others are saying in their documentation:
https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/docs/semantics.md#caching

File: src/client/dfuse/dfuse_core.c
Comment (ashleypittman) [inline src/client/dfuse/dfuse_core.c:685]: A more cautious approach would to to set this to 0, at least until we better understand not to handle negative dentry caching on unlink.

File: src/client/dfuse/dfuse_core.c
Comment (ashleypittman) [inline src/client/dfuse/dfuse_core.c:700]: As L685.  I'll merge these two code-blocks into a "dfuse_set_cache_defaults()" function on repush to make this clear.

File: src/client/dfuse/ops/rename.c
Comment (ashleypittman) [inline src/client/dfuse/ops/rename.c:46]: I'm unhappy with this as it seems there's a race condition where the kernel or other processes might end up doing a lookup which would result in ENOENT before we've had a chance to invalidate the dentry.  I'm working on a standalone reproducer which I'll send to the fuse maintainers to seek their opinions.

File: src/client/dfuse/dfuse_core.c
Comment (ashleypittman) [reply src/client/dfuse/dfuse_core.c:685]: I've set it to 1

File: src/client/dfuse/dfuse_core.c
Comment (ashleypittman) [reply src/client/dfuse/dfuse_core.c:700]: Done.

File: src/tests/ftest/util/ior_test_base.py
Comment (daltonbohning) [inline src/tests/ftest/util/ior_test_base.py:40]: I'm not sure this works, since in dfuse_test_base.py, a new Dfuse() command is created each time it runs, and I don't think IorTestBase.disable_caching is taken into account.
https://github.com/daos-stack/daos/blob/26a478afa6252c99bfa89cee17b0414a28629a63/src/tests/ftest/util/dfuse_test_base.py#L39-L48

**Review by daltonbohning — COMMENTED** [review]
On a previous run, it looks like some Centos8 datamover tests failed to start dfuse. FWIW, Danielle saw something similar the other day on a different PR, so I don't think this PR is causing that.

File: src/client/dfuse/dfuse_core.c
Comment (ashleypittman) [reply src/client/dfuse/dfuse_core.c:62]: Will fix if I need to push.

File: src/client/dfuse/dfuse_main.c
Comment (ashleypittman) [reply src/client/dfuse/dfuse_main.c:232]: We've never encouraged caching before and the behaviour is completely different after this PR so no, I don't think in this case that we need to.

File: doc/user/posix.md
Comment (ashleypittman) [reply doc/user/posix.md:197]: Yes, that's a good idea.  I'll encapsulate it in a ticket once this lands.

File: src/client/dfuse/ops/open.c
Comment (ashleypittman) [reply src/client/dfuse/ops/open.c:48]: Yes, I was going for clarity over brevity ;)

File: src/tests/ftest/util/ior_test_base.py
Comment (ashleypittman) [reply src/tests/ftest/util/ior_test_base.py:40]: You may well be right, we'll have to see.  An older version of this PR failed IOR and CI hasn't been stable enough to run this code yet.  I was hoping that object inheritance would take care of this but I'll admit to not having a firm grasp on how ftest works.  We should have results in an hour or so and we can take a close look then.

File: src/client/dfuse/ops/open.c
Comment (daltonbohning) [reply src/client/dfuse/ops/open.c:48]: No worries. Just making sure I understood what was happening :)

File: src/tests/ftest/util/ior_test_base.py
Comment (daltonbohning) [reply src/tests/ftest/util/ior_test_base.py:40]: I do think it would be ideal if the ior results work "as-is". I.e. with the default of caching enabled.

Otherwise, if we do need to disable...
The problem here is that the Dfuse command itself is not part of the test "base" inheritance chain. If you want IorTestBase.disable_caching to be taken into account automatically for all tests inheriting IorTestBase, I _**think**_ we might need something like:
```
self.dfuse = Dfuse(hosts, self.tmp)
# Handle default caching behavior
if self.disable_caching:
    self.dfuse.disable_caching = True
# Now get all params from config (.yml), including disable_caching
self.dfuse.get_params(self)
```
And to be more clear in DfuseTestBase, we should do this in `DfuseTestBase.__init__`
```
def __init__(...):
    # By default, do not disable and get the value from config
    self.disable_caching = False
```
That way IorTestBase.disable_caching is setting the inherited DfuseTestBase.disable_caching.
And when DfuseTestBase runs dfuse, it will first set DfuseTestBase.disable_caching, and then optionally override that with the value from the test config (.yml).
The end result is:
- By default, dfuse enables caching.
- By default, any test inheriting IorTestBase will disable caching
- Individual test cases can enable caching by putting this in the test config (.yml):
  ```
  dfuse:
    disable_caching: false
  ```

But yes, we can wait until the results are in to see if this is necessary.

File: src/tests/ftest/ior/ior_small.yaml
Comment (daltonbohning) [inline src/tests/ftest/ior/ior_small.yaml:35]: We should probably remove this since a newer commit adds `disable_caching` lower in this config

Comment (ashleypittman) [reply src/tests/ftest/ior/ior_small.yaml:35]: I thought I had, just about to re-push this so I'll do it next time.

Comment (daltonbohning) [reply src/tests/ftest/ior/ior_small.yaml]: Oh, you're right, you did. GitHub was lying to me :(

File: src/tests/ftest/util/ior_test_base.py
Comment (ashleypittman) [reply src/tests/ftest/util/ior_test_base.py]: > I do think it would be ideal if the ior results work "as-is". I.e. with the default of caching enabled.

It would, if we only ran IOR on a single node.

> Otherwise, if we do need to disable...
> The problem here is that the Dfuse command itself is not part of the test "base" inheritance chain. If you want IorTestBase.disable_caching to be taken into account automatically for all tests inheriting IorTestBase, I _**think**_ we might need something like:
> 
> ```
> self.dfuse = Dfuse(hosts, self.tmp)
> # Handle default caching behavior
> if self.disable_caching:
>     self.dfuse.disable_caching = True
> # Now get all params from config (.yml), including disable_caching
> self.dfuse.get_params(self)
> ```

This is the bit I was missing.  What I've done is just added the option, and set it in the yaml for all tests where it's needed.  I've pushed this, and triggered a daily_regression build to be sure but the previous one only hit ior and fio failures, and I've already tested both of those since.

Comment (daltonbohning) [reply src/tests/ftest/util/ior_test_base.py]: That is probably the simplest approach, thought I'm not sure if other tests do or do not need to be ran.

**Review by daltonbohning — COMMENTED** [review]
I'd recommend running with:
`Test-tag: pr datamover`
To run the datamover tests as well. Particularly the `datamover,hw,large` tests, since some of those use dfuse with 16 processes on a structure with 40k files + 40k dirs.
(`Features` tag doesn't work here because they are full_regression test cases, so `Test-tag` is needed.)

Assuming all relevant tests are passing, LGTM.

**Review by johannlombardi — DISMISSED** [review]

**Comment by johannlombardi** [comment]
It looks like the fit issues are relevant here. Since those are single-node tests, i don't see why they would fail with caching.

**Comment by ashleypittman** [comment]
> It looks like the fit issues are relevant here. Since those are single-node tests, i don't see why they would fail with caching.

fio tests are running with caching disabled anyway.  I'm investigating.

**Comment by ashleypittman** [comment]
> I'd recommend running with:
> `Test-tag: pr datamover`
> To run the datamover tests as well. Particularly the `datamover,hw,large` tests, since some of those use dfuse with 16 processes on a structure with 40k files + 40k dirs.
> (`Features` tag doesn't work here because they are full_regression test cases, so `Test-tag` is needed.)
> 
> Assuming all relevant tests are passing, LGTM.

Merged and pushed with this tag whilst I investigate the fio failures.

File: utils/node_local_test.py
Comment (ashleypittman) [reply utils/node_local_test.py:2220]: Will fix, but there's a high priority test run in progress currently.

**Comment by ashleypittman** [comment]
This passed a daily_regression overnight, can I request re-reviews please.

**Comment by johannlombardi** [comment]
Thanks Ashley. @mchaarawi & @daltonbohning, i would appreciate your prompt review. Thanks in advance.

File: src/client/dfuse/dfuse_core.c
Comment (mchaarawi) [inline src/client/dfuse/dfuse_core.c:523]: we could get all attributes in a single call, no?

**Review by mchaarawi — APPROVED** [review]

File: src/client/dfuse/dfuse_core.c
Comment (ashleypittman) [reply src/client/dfuse/dfuse_core.c:523]: I don't think we can yet, but @bfaccini is working on it.  At the minute we could only get all attributes in a single call if they all exist. https://jira.hpdd.intel.com/browse/DAOS-6710

We'll update this code once the feature is available.

**Review by daltonbohning — APPROVED** [review]
It doesn't look like the commit with "Test-tag: pr datamover" completed the test stages (restarted/new commit). It's _**probably**_ okay, and running some tests locally, `dcp` seems okay. But if you need to make another commit, the datamover tests are a good sanity check. Specifically the "datamover,hw,large" tests, since some of them use 16 MPI processes (5 client nodes) on a large dfuse file/directory.

**Comment by ashleypittman** [comment]
> It doesn't look like the commit with "Test-tag: pr datamover" completed the test stages (restarted/new commit). It's _**probably**_ okay, and running some tests locally, `dcp` seems okay. But if you need to make another commit, the datamover tests are a good sanity check. Specifically the "datamover,hw,large" tests, since some of them use 16 MPI processes (5 client nodes) on a large dfuse file/directory.

Yes you're right.  I thought this was build #52 but it seems not, and there were some issues with Unit Test.  I've created https://jira.hpdd.intel.com/browse/DAOS-7621 to create a dfuse test-tag so we can catch all affected tests in future.

I'd like to land this PR as-is given it's green and with approvals (I didn't realise github-actions failed overnight but I've just retriggered it), but I'm going to work on another PR to improve and expand on the documentation and I'll addresses your datamover concerns there.

File: doc/user/posix.md
Comment (ashleypittman) [reply doc/user/posix.md:197]: https://jira.hpdd.intel.com/browse/DAOS-7652

File: doc/user/posix.md
Comment (ashleypittman) [reply doc/user/posix.md:228]: I'm doing another pass on this, and will add what we discovered on aio at the same time.

**Pull request merged**
